lib/awful/completion.lua

```diff
@@ -154,10 +161,25 @@ function completion.shell(command, cur_pos, ncomp, shell)
         while true do
             local line = c:read("*line")
             if not line then break end
-            if os.execute("test -d " .. string.format('%q', line)) == 0 then
-                line = line .. "/"
+            if comptype == "command" then
+                if str_starts(line, ".") or str_starts(line, "/") then
+                    if gfs.file_executable(line) then
+                        table.insert(output, bash_escape(line))
+                    end
+                else
+                    for path in string.gmatch(os.getenv('PATH'), "[^:]+") do
+                        if gfs.file_executable(path .. "/" .. line) then
+                            table.insert(output, bash_escape(line))
+                            break
+                        end
+                    end
+                end
+            end
+            if gfs.is_dir(line) then
+                table.insert(output, bash_escape(line .. "/"))
+            elseif comptype == "file" then
+                table.insert(output, bash_escape(line))
             end
-            table.insert(output, bash_escape(line))
```

Comment:
Sadly, bashcomp does not provide everything ... The main reason I use this overly complex construct is because we cannot know if the current item is "just" a directory or if there is also an executable somewhere in the path.

Regarding aliases: With `-A command`, we already exclude aliases and functions (at least with bash). If you want to complete aliases, too, then the bash-code needs to be updated.

I think, the whole confusion arises from the requirement to complete local directories, which is by the way not even consistent in bash itself. A directory search for example is only done if absolutely no other match has been found. And if a directory is found, a slash is only added if there is no command/executable/alias with the same name. So maybe we should first work out, how much of the (sometimes strange) behavior we really want to re-produce. Wouldn't commands (and aliases for this sake) be enough? Is it really necessary to add directories to the completion results? If we just want to complete commands, then the executable check as well as the directory check can be ditched altogether in favor of a slightly modified completion command.